### PR TITLE
Floria: Implement run context and enable recursive calls

### PR DESCRIPTION
**Pull request by simonlechner**
This PR depends on #627 and implements the run context which allows for recursive calls from the interpreter.

### CODE REVIEW

File: go/processor/floria/processor.go [hunk 1]
```diff
@@ -23,6 +23,8 @@ const (
 	TxDataZeroGasEIP2028      = 4
 	TxAccessListAddressGas    = 2400
 	TxAccessListStorageKeyGas = 1900
+
+	CallCreateDepth = 1024 // Maximum depth of call/create stack.
```
Comment (HerbertJordan) [inline go/processor/floria/processor.go:27]: nit: `CallCreateDepth` => `MaxRecursiveDepth` or `MaxCallNestingDepth`

File: go/processor/floria/processor.go [hunk 2]
```diff
@@ -93,6 +112,14 @@ func (p *processor) Run(
 	}, nil
 }
 
+type runContext struct {
+	tosca.TransactionContext
+	interpreter           tosca.Interpreter
+	blockParams           tosca.BlockParameters
+	transactionParameters tosca.TransactionParameters
```
Comment (HerbertJordan) [inline go/processor/floria/processor.go:119]: nit: would be nice to be consistent ... either both are `Params` or `Parameters`

File: go/processor/floria/calls.go [hunk 3]
```diff
@@ -16,62 +16,70 @@ import (
 	"github.com/Fantom-foundation/Tosca/go/tosca"
 )
 
-func call(interpreter tosca.Interpreter, transaction tosca.Transaction, context tosca.TransactionContext, gas tosca.Gas) (tosca.Result, error) {
+func (r runContext) Call(kind tosca.CallKind, parameters tosca.CallParameters) (tosca.CallResult, error) {
```
Comment (HerbertJordan) [inline go/processor/floria/calls.go:19]: It is surprising that the `runContext` is defined in the `processor.go` but its main only member function `Call` is defined here. I think the type definition should be moved here and the file may be renamed from `call.go` to `run.go` or `run_context.go`. Wdyt?

File: go/processor/floria/calls.go [hunk 4]
```diff
@@ -16,62 +16,70 @@ import (
...
 		return fmt.Errorf("insufficient balance: %v < %v", senderBalance, transferValue)
 	}
 
 	senderBalance = tosca.Sub(senderBalance, transferValue)
-	context.SetBalance(transaction.Sender, senderBalance)
+	context.SetBalance(sender, senderBalance)
 
-	receiverBalance := context.GetBalance(*transaction.Recipient)
+	receiverBalance := context.GetBalance(recipient)
 	receiverBalance = tosca.Add(receiverBalance, transferValue)
```
Comment (HerbertJordan) [inline go/processor/floria/calls.go:81]: If we want to play it super-safe we would have to check that adding the transferred value is not leading to an overflow at the receiver. In such a case, we would also let the transfer fail.

File: go/processor/floria/calls.go [hunk 5]
```diff
@@ -16,62 +16,70 @@ import (
...
-		Recipient: *transaction.Recipient,
-		Sender:    transaction.Sender,
-		Input:     transaction.Input,
-		Value:     transaction.Value,
-		CodeHash:  &codeHash,
-		Code:      code,
+		BlockParameters:       r.blockParams,
+		TransactionParameters: r.transactionParameters,
+		Context:               r,
+		Kind:                  tosca.Call,
```
Comment (HerbertJordan) [inline go/processor/floria/calls.go:34]: I guess the `Kind` should be `kind` ... but this will get unearthed when testing different call types.

**Review by HerbertJordan — CHANGES_REQUESTED** [review]
The architecture looks good, just a few requests regarding code organization. Please have a look.

File: go/processor/floria/calls.go [hunk 6]
```diff
@@ -16,62 +16,70 @@ import (
...
+) error {
+	senderBalance := context.GetBalance(sender)
+	transferValue := value
 	if senderBalance.Cmp(transferValue) < 0 {
 		return fmt.Errorf("insufficient balance: %v < %v", senderBalance, transferValue)
 	}
 
 	senderBalance = tosca.Sub(senderBalance, transferValue)
-	context.SetBalance(transaction.Sender, senderBalance)
+	context.SetBalance(sender, senderBalance)
```
Comment (LuisPH3) [inline go/processor/floria/calls.go:78]: is negative balance allowed?

**Review by LuisPH3 — DISMISSED** [review]

File: go/processor/floria/calls.go (hunk 6 above)
Comment (simonlechner) [reply go/processor/floria/calls.go:78]: No, there is a check whether the sender has sufficient balance.

File: go/processor/floria/calls.go (hunk 3 above)
Comment (simonlechner) [reply go/processor/floria/calls.go:19]: Right, I moved it to the calls file and renamed it to run_context.go

File: go/processor/floria/calls.go (hunk 4 above)
Comment (simonlechner) [reply go/processor/floria/calls.go:81]: Right, this has been missed in the last PR were this function was introduced. I restructured the whole function and added more unit tests.

**Pull request merged**